### ISSUE TITLE
Feature request: universal REPL

### ISSUE THREAD

**Issue by benjaminrich** [helix-editor/helix#2806]
I'm pretty impressed with this:

https://gitlab.com/Screwtapello/kakoune-repl-buffer

(also, watch the video here: https://asciinema.org/a/379916)

If something like this could be added to helix, it could potentially change my life. I know some other users are looking for this kind of feature as well (#1710).

Any thoughts?

**Comment by the-mikedavis** [comment]
I think this would be a cool feature that could fit into core. At first I thought it might be better as a plugin since it isn't based on some hypothetical REPL standard (something like LSP for REPL) but it looks like the plugin is just interfacing with stdin/stdout of the repl program which seems straightforward and consistent.

There are some rough edges to that plugin that I think could be designed around: the `:prompt` based capabilities at the end of the asciicast effectively turn kakoune into a line-editor. That fine for simple lines on a REPL but it's limiting for larger code fragments and it would be ideal to be able to use the full buffer editing power of Helix instead of prompts.

I don't have any ideas to get the ball rolling but it could be smart to canvas repl popular repl plugins from vim/emacs/etc and source ideas there.

**Comment by benjaminrich** [comment]
@the-mikedavis, thanks for your response.

To be clear, I'm not suggesting to re-implement that plugin exactly (I agree with you that it is rough and about the `:prompt` functionality, etc). I was just trying to point to an example that goes in the right direction. Ideally, I would like it to be a nice, well integrated feature with easy, consistent key bindings. I am thinking that it could be useful to have a `REPL` minor mode (with a sticky option) to "send code to REPL" and that sort of thing (admittedly I haven't quite thought this through, it's just an idea).

I mostly do work in R, and a bit of julia. For R, there is the very nice `nvim-R` plugin, that has REPL functionality, with lots of features. Unfortunately, it only works for R and can't be used for other interactive languages (like julia for instance), where you would have to rely on something like `slime.vim`, thus creating a major inconsistency in the workflows. That is the situation I would like to rectify.

[Sidenote: there is an official julia plugin for R, but it doesn't have any REPL features. It does however allow for nice entry of some Unicode characters, which can also be used in non-julia files, and is a useful feature in its own right. But that is a topic for a different discussion.]

Also, I am very aware of the "kitchen sink" and "bloat" concern. Indeed, this functionality goes beyond simple editing, but for interactive languages (and there are many) it can be considered "essential" (I would never go back to copy-pasting code into a different terminal/window). Then there are external tools like tmux that *could* be leveraged, but here we have the problem of catering to all kinds of users with their vastly heterogeneous environments, and so it really is better not to rely on the presence of external tools (approaches like that also always feel kind of kludgy). Obviously, there may be users that do not consider this feature important, but I hope there are enough who do.

Plugins create a plethora of fragmentation problems (as exemplified by `nvim-R` above), which is why I was *very* excited to hear that you feel this could fit into core. Consistent, well designed and well integrated with the rest of helix - that is the best solution. It could be a "killer" feature that distinguishes helix from other modal editors.

[Other sidenote: SpaceVim has built-in REPL functionality and defines some consistent bindings across different languages (layers in SpaceVim). Unfortunately they are pretty cumbersome because they require like 4 key presses, which is really too much].

Thanks, and sorry for the long post.

**Comment by dgkf** [comment]
Just wanted to add to the reference material. I'm most familiar with [`vim`/`neovim`'s Terminal mode](https://vimhelp.org/terminal.txt.html#terminal-use), which is a full interactive terminal emulator rendered in a window. I use a plugin ([`vim-slime`](https://github.com/jpalardy/vim-slime)) to map a keybind to send code to the terminal. This is modeled after [`emacs`' slime mode](https://www.emacswiki.org/emacs/SlimeMode). 

[![image](https://user-images.githubusercontent.com/18220321/186534361-768bb8f2-bd35-4382-aad6-383441e97448.png)](https://asciinema.org/a/hqFTlEuJ0gvPazrpi91dJlUU2)
It's not pretty (I disabled most other plugins), but hopefully it's clear how it works.

As far as use cases go, I'm also coming from an R-for-analytics background. In the domain of scientific programming, interactive execution in a REPL is a very central part of development.

**Comment by riwsky** [comment]
> Something like LSP for REPL

the [Jupyter protocol](https://jupyter-client.readthedocs.io/en/latest/) fits this description, with [plenty](https://github.com/jupyter/jupyter/wiki/Jupyter-kernels) of implementations by this point

**Comment by benjaminrich** [comment]
> > Something like LSP for REPL
> 
> the [Jupyter protocol](https://jupyter-client.readthedocs.io/en/latest/) fits this description, with [plenty](https://github.com/jupyter/jupyter/wiki/Jupyter-kernels) of implementations by this point
> 

I always thought of Jupyter as something web based, but maybe I don't understand it very well. Could it work in a console editor like helix?

**Comment by riwsky** [comment]
Yeah.

While the web-based notebook is by now Jupyter's calling card, the project started life as a python console REPL. The project still supports that console REPL, along with a few other frontends (a QT REPL, alternative web UIs, etc). They all talk to backends over a [ZeroMQ-based protocol](https://jupyter-client.readthedocs.io/en/latest/index.html).

Granted: these first-party UIs represent the overwhelming majority of Jupyter use, so their "first among equals" status would be even stronger than VSCode's is for LSP and DAP. That hasn't stopped people from trying, though:

[jupyter-vim](https://github.com/jupyter-vim/jupyter-vim):
> A two-way integration between Vim and Jupyter. Develop code on a Jupyter notebook without leaving the terminal. Send lines from Vim to a jupyter qtconsole. Have a MATLAB-like "cell-mode".

[emacs-jupyter](https://github.com/nnicandro/emacs-jupyter):
> An interface to communicate with Jupyter kernels in Emacs.

[(VS Code) Python Interactive window](https://code.visualstudio.com/docs/python/jupyter-support-py):

> [Jupyter](https://jupyter-notebook.readthedocs.io/en/latest/) (formerly IPython Notebook) is an open-source project that lets you easily combine Markdown text and executable Python source code on one canvas called a notebook. Visual Studio Code supports working with [Jupyter Notebooks natively](https://code.visualstudio.com/docs/datascience/jupyter-notebooks), as well as through Python code files.

[(PyCharm) Scientific mode](https://www.jetbrains.com/help/pycharm/matplotlib-support.html#sm)
> Scientific mode in PyCharm provides support for interactive scientific computing and data visualization.

I haven't been a regular user of editor built-in Jupyter support since [a decade ago](https://github.com/ivanov/vim-ipython)—but from what I can tell, there are a few features that have emerged as conventional to support:

1. automatically launch a backend kernel if one isn't running
2. run selected code
3. separate code into cells with a magic comment (typically `# %%`), and make it easy to run entire cells at a time
4. automatically open a console buffer in a split
5. debugger support

helix already has many discussions in flight for 4 and 5, but for these other features:

1. Helix already launches LSPs automatically, so this seems uncontroversial.
2. This generalizes the existing shell-specific behaviors in `|` and friends, so the feature itself doesn't seem out of place. Evaluating code tends to be a pretty common action in read-eval-print-loops, though, so you'd want an easily-accessible keymap (ideally involving Enter)—and it's not clear whether dedicating such valuable real estate to REPLs makes sense for what is primarily a text editor.
3.  This is a little uglier. The way other tools do it does not fit helix's select-then-act paradigm—and even if helix made it easy to select within a `# %%` cell (e.g. with multi-character surround, or defining text objects for them and supporting syntax surround), it still smells bad to make users change code to help the editor. It's clever, though, and it's not obvious what to do instead—maybe helix could track which selections on the jumplist were created explicitly (i.e. with `save_selection`), and support those as a text object in match mode? That'd give you, say, `m<Ctrl-s>` for "select current cell", which doesn't seem that bad.

**Comment by benjaminrich** [comment]
@riwsky Thanks a lot for your detailed response! In that case, it does sounds like a good approach. I agree with your comment about keybindings. Ideally, you want to be able to "spam" the `evaluate-current-and-move-to-next` command with something like ctrl-enter or shift-enter (i.e., hold down one modifier and press one key, not multiple keystrokes).

**Comment by danjenson** [comment]
I'm trying to create a temporary workaround for this using tmux with the following script:
```bash
#!/bin/bash
# TODO(danj): take argument specifying language/REPL to start
tmux set-buffer "$(xclip -o -sel clipboard)"
tmux paste-buffer -t "{down-of}" 2> /dev/null
if [ $? -ne 0 ]; then
  tmux split-window -v -p 35
  # TODO(danj): start REPL here
  tmux paste-buffer
  tmux select-pane -t "{up-of}"
fi
```
However, I can't seem to extract the file type from helix to start the appropriate REPL. I suppose I could do something hacky by inspecting processes and starting a REPL based on that, but it would be nice if you could do something like `:sh my_script.sh (file-type)` where `file-type` resolves to the current file type OR if you can specify custom shortcuts for each language defined in languages.toml

**Comment by sourproton** [comment]
> Consistent, well designed and well integrated with the rest of helix - that is the best solution. It could be a "killer" feature that distinguishes helix from other modal editors.

I definitely agree with that! 

I just want to share my thoughts on the "launching a REPL if one doesn't exist":

What I like best about my current neovim/slime setup is that I'm free to easily setup the REPL I want, how I want it, in a terminal window or a tmux pane. The plugin simply sends the code to the target REPL, with the convenience of setting magic comments like `# %%` to run the entire cell with a single shortcut.

I don't think that the Helix REPL feature should setup a REPL for the user if one isn't active. With different languages requiring different setups, this level of abstraction can get difficult to implement very quickly. What if there are different active REPLs? How to choose the environment and working directory of the REPL?

The user can easily launch their REPL in the language they want, in the environment they want. Sourcing a Python environment or setting a Julia project is easy and I think that a user who needs REPL integration knows how to do it.

I think the REPL integration in Helix should focus on implementing a convenient way to target an existing REPL and reliably send the selected text or the current cell to that REPL.

Some people use different magic comments for their cells, so I think this could be a configuration option, that defaults to "# %%".

**Comment by sourproton** [comment]
Another benefit of the REPL being outside of the editor, on top of being easier to setup the desired environment, is that the REPL can be on a ssh connection. I myself run the editor locally and pass the code to a remote REPL on a server.

**Comment by rapasite** [comment]
Just adding some cool reference material.
Inline result is a way of coding that is so satisfactory and productive.

-[https://git.sr.ht/~detegr/nvim-bqn](url)
This way of doing it feel the best for me.

-[https://github.com/mrossinek/deuterium](url)
Deuterium is a vim-implementation of the popular hydrogen plugin for Atom.

-[https://github.com/dccsillag/magma-nvim](url)
Magma is a NeoVim plugin for running code interactively with Jupyter.

Hopefully someone talented can implement something like that.
That would be so nice to have for data science workflow.

**Comment by theophilusx** [comment]
I'm very impressed with Helix and I would also find the ability to have basic repl integration extremely useful. However, I would suggest this will take considerable design effort to get right. Often I am a great believer in just building something and then refining it based on usage and feedback. However, in other systems I've used which adopted this approach with respect to interfacing with repls and sub-processes generally, the result has tended to be overly complex with way too many options and points of confusion with respect to correct behaviour in various corner cases. 

The biggest source of problems seems to centre around definitions of output, return values, error codes and handling of stdout and stderr. Things become more complicated because different languages have different semantics here - some languages always return a value (Clojure, FP langs) , some only when the code explicitly does so (Python, many OOP langs) and some rely on stdout/sterr and process return codes (shells, CLI tools). Sometimes you may want the value returned by the repl or sub-process and sometimes you might want whatever was sent to stdout and/or stderr and sometimes you only care about the exit code. Ability to specify this in a clean and concise m,anner will be critical to a solid implementation. 

The other things to possibly think about would be session management. Sometimes, you want all the code evaluation to happen within the same session context and other times, you mgiht want the code executed in a clean context or you may just want an easy way to clear the context. 

The challenge is in how to provide this functionality in a flexible manner which is not complicated and which doesn't detract from its big strength - clarity and simplicity without sacrificing functionality. When I compare my neovim and helix configuration where I have largely equivalent functionality, the benefits of Helix are obvious. The ability to interface with a repl needs to maintain that clarity and simplicity if possible.

**Comment by jerabaul29** [comment]
This is really awesome and looks very promising. As far as I am concerned (and I am likely not alone, many people with some data science / ml / scientific programming background around me are in the same case), having a way to use functionality 'a la jupyter(lab)' is what will make me take the step to move to Helix.

Wondering: even if a native / official solution is still in the pipelines, would it be possible to have a repo or folder on this repo somewhere, either Helix or wider community maintained, that:

- list different alternatives available as of today for getting an experience 'a la jupyter(lab)'

- for (as many as possible of) these, provide a dedicated detailed tutorial or points to one, so that anyone with minimal technical proficiency is able to set up such an environment and use it

?

The idea would be that, in the meantime waiting for 'the ultimate Helix native solution', there may be 'perfectly good enough' temporary solutions, and that making these easy to use, and nurturing an ecosystem of users, and helping users migrate to better solutions as these are developed, would be very helpful and help the user community grow.

**Comment by jonas-w** [comment]
@jerabaul29 I don't know if this fits your needs, but I use [euporie](https://github.com/joouha/euporie) to edit and run jupyter python notebooks in my console, it also has a vim mode (i wish it had a helix mode). But I agree it would be really awesome to have something like this in helix itself, with more familiar keybindings and faster iteration while writing code.

**Comment by sourproton** [comment]
I would like to share a setup that is almost perfectly working for me. The only downside is that it relies on a macro that I can't seem to find a way to wrap in a bind.

The coding setup is based on tmux, with code on the left and a REPL on the right. The target REPL can be anywhere though, even in another tmux session, you just need to modify the `-t` flags in the tmux commands.

So, inspired by @dgkf, I added
```
[keys.normal.'\']
space = [":pipe-to tmux load-buffer - \\; paste-buffer -dpr -t right", "tmux send-keys -t right Enter"]
```
to my Helix config. Once we select the text we want to send to the REPL with Helix's select mode, `:pipe-to tmux load-buffer - \\` loads a tmux buffer with that text, and `paste-buffer -dpr -t right` pastes it to the pane in the right. Notice the flags: they are not mandatory, but that way it works the best for me, specially the `-p` flag with pastes the text in bracketed mode, which allows better handling of blank lines in ipython, for example. Because of the bracketed paste, `tmux send-keys -t right Enter` is needed to send a `Enter` keystroke to the REPL, to execute the pasted code, otherwise some REPLs (like ipython) wil just paste the code and not run it.

That would be enough for many, but I find it counter-productive to have the need to select the text to be send every time, specially because in my REPL workflow I work with cell headers in my scripts, like `# %%`. Here's an example code:

```
# %% cell 1
print("foo")

# %% cell 2
def bar(x):
    # compute y
    y = 2 * x

    return y

print(bar(1))

# %% cell 3
print("baz")
```

The ideal scenario for me is to be anywhere in the file and, with a keystroke, execute the whole cell containing the cursor and go to the next cell. I would like to have this bound to a key, but could only achieve this with a macro. So, for now, I need to re-type the macro every time I start a new Helix instance. Inspired by [this reddit comment](https://www.reddit.com/r/HelixEditor/comments/13i527h/comment/jk8r6et/?utm_source=share&utm_medium=web2x&context=3), I arrived at

```Q/# %%<RET>O#<ESC>Nvg.a<BS><BS><ESC>\<SPACE>uuQ```

It seems long, but the logic is straight forward:
1. `Q` to start recording the macro
2. `/# %%<RET>` searches for the next cell header
3. `O#<ESC>` puts a character at the end of the cell we and to execute. This modifies the file and, once the cursor is at the start of the cell, we can use this to navigate to the end of the cell
4. `N` goes to the start of the cell. It works because of `2.`
5. `v` to enter select mode
6. `g.` to select until the last modification, that is, the `#` inserted at `3.`
7. `a<BS><BS><ESC>` to remove the `#`
8. `\<SPACE>` is the bind that calls  `[":pipe-to tmux load-buffer - \\; paste-buffer -dpr -t right", "tmux send-keys -t right Enter"]`
9. `uu` is needed because we added and removed the `#` mark, thus modifying the file. Undoing that clears the modification history and, luckily, puts the cursor at the start of the next cell
10. `Q` to stop recording the macro

Now just type `q` to execute the macro and jump to the next cell. To execute it, just type `q` again.

Anyway I'm 90% happy with this because I still need to setup the macro every time. If someone has a better suggestion it would be very welcome.

**Comment by sourproton** [comment]
I would like to share a setup that is perfectly working for me. 

## TLDR
1. Add this to your Helix config:
```
[keys.normal.'\']
space = [":pipe-to tmux load-buffer - \\; paste-buffer -dpr -t '{right-of}'", ":sh tmux send-keys -t '{right-of}' Enter"]
'\' = [":sh tmux send-keys Q/# Space %% Enter O# Escape Escape Nvg.a BSpace BSpace Escape Escape '\\' Space uuQ"]
# the above is the quivalent of putting, in .tmux.conf: send-keys Q/#\s\%\%\rO#\e\eNvg.a\b\b\e\e\\\suuQ
```
2. Have a tmux session with a REPL in the pane directly to the right of the pane containing Helix
3. Have the cursor within a cell, delimited with the `# %%` header, call the `\\` bind to run the first cell you want to execute (not necessarily the first cell of the notebook). This records a macro. Then use `q` to run any other cell, provided the cursor is within its limits. `n` and `N` will also jump to the next and previous cells.
4. Alternatively, select the text you want to send to your REPL and use the `\-Space` bind to send it.

Here's the result, with ipython and Julia REPLs:

![output](https://github.com/helix-editor/helix/assets/66533348/d7b6a04f-db4b-4cc3-bd8c-f7ee6c69cd44)

## Explanation

The coding setup is based on tmux, with Helix containing the code on a pane a REPL on the pane directly to the right. The target REPL can be anywhere though, even in another tmux session, you just need to modify the `-t` flags in the tmux commands.

So, inspired by @dgkf, I added
```
[keys.normal.'\']
space = [":pipe-to tmux load-buffer - \\; paste-buffer -dpr -t '{right-of}'", ":sh tmux send-keys -t '{right-of}' Enter"]
```
to my Helix config. Once we select the text we want to send to the REPL with Helix's select mode, `:pipe-to tmux load-buffer - \\` loads a tmux buffer with that text, and `paste-buffer -dpr -t '{right-of}` pastes it to the pane in the right. Notice the flags `-dpr`: they are not mandatory, but that way it works the best for me, specially the `-p` flag that pastes the text in bracketed mode, which allows better handling of blank lines and indentation in ipython, for example. Because of the bracketed paste, `:sh tmux send-keys -t '{right-of}' Enter` is needed to send a `Enter` keystroke to the REPL, to execute the pasted code, otherwise some REPLs (like ipython) will just paste the code and not run it.

That would be enough for many, but I find it counter-productive to have the need to select the text to be sent every time, specially because in my REPL workflow I work with cell headers in my scripts, like `# %%`. Here's an example code:

```
# %% cell 1
print("foo")

# %% cell 2
def bar(x):
    # compute y
    y = 2 * x

    return y

print(bar(1))

# %% cell 3
print("baz")
```

The ideal scenario for me is to be anywhere in the file and, with a keystroke, execute the whole cell containing the cursor and go to the next cell. I would like to have this bound to a key, but could only achieve this with a macro. So, for now, I need to re-type the macro every time I start a new Helix instance. Inspired by [this reddit comment](https://www.reddit.com/r/HelixEditor/comments/13i527h/comment/jk8r6et/?utm_source=share&utm_medium=web2x&context=3), I arrived at

```Q/# %%<RET>O#<ESC>Nvg.a<BS><BS><ESC>\<SPACE>uuQ```

It seems long, but the logic is straight forward:

0. Have the cursor anywhere in the cell
1. `Q` to start recording the macro
4. `/# %%<RET>` searches for the next cell header (notice the space between `#` and `%%`, respecting the cell header)
5. `O#<ESC>` puts a character at the end of the cell we and to execute. This modifies the file and, once the cursor is at the start of the cell, we can use this to navigate to the end of the cell
6. `N` goes to the start of the cell. It works because of `2.`
7. `v` to enter select mode
8. `g.` to select until the last modification, that is, the `#` inserted at `3.`
9. `a<BS><BS><ESC>` to remove the `#`
10. `\<SPACE>` is the bind that calls  `[":pipe-to tmux load-buffer - \\; paste-buffer -dpr -t '{right-of}'", ":sh tmux send-keys -t '{right-of}' Enter"]`. Modify this command to send the text to other terminal multiplexers.
11. `uu` is needed because of the added and after removed `#` mark, that modified the file and polluted the change history. Undoing that clears the history and, luckily, puts the cursor at the start of the next cell
12. `Q` to stop recording the macro

Now just type `q` to execute the macro and jump to the next cell. To execute it, just type `q` again. The cursor doesn't need to be at the beginning of the cell, it can be anywhere inside it. Also, because of `2.`, `n` and `N` will jump no the next and previous cell, too.

I was 90% happy with this because I still needed to setup the macro every time and it doesn't work for the last cell.

To mitigate the last cell problem, a simple solution is to have
```
# %% end
```
at the end of the file and `n` or `N` to jump it.

Finally, to avoid needing to re-type the long macro every time, a solution is to add another bind to Helix, that will call the `send-keys` tmux cli function to type the macro by itself. The resulting Helix config is:

```
[keys.normal.'\']
space = [":pipe-to tmux load-buffer - \\; paste-buffer -dpr -t '{right-of}'", ":sh tmux send-keys -t '{right-of}' Enter"]
'\' = [":sh tmux send-keys Q/# Space %% Enter O# Escape Escape Nvg.a BSpace BSpace Escape Escape '\\' Space uuQ"]
# the above is the quivalent of putting, in .tmux.conf: send-keys Q/#\s\%\%\rO#\e\eNvg.a\b\b\e\e\\\suuQ
```

Double `Escape`s are needed, I think because of some kind of delay. With a single `Escape`, Helix doesn't quit insert mode before the next `send-keys` command.

So now I can `\\` to record the macro automatically and then use `q` to run it again as many times as needed.

**Comment by kshitijsachan** [comment]
@sourproton Thanks for sharing! Is there some way to adapt that script to pipe to Jupyter lab? I use Helix on a remote machine and would like to be able to display interactive plots (which won't work in tmux).

**Comment by aberges-grd** [comment]
I wanted to barge in real quick to say that the "cell division standard" in buffers could be set as the "Form Feed" character, inserted through Ctrl+L in e.g. Emacs, which is a standard of the ASCII code for page break. As such, it perfectly delimits pieces of code larger than "blocks" but shorter than "buffers", would not collide with any other character nor interfere with other editors that do not use it.

**Comment by astro-ray** [comment]
> @jerabaul29 I don't know if this fits your needs, but I use [euporie](https://github.com/joouha/euporie) to edit and run jupyter python notebooks in my console, it also has a vim mode (i wish it had a helix mode). But I agree it would be really awesome to have something like this in helix itself, with more familiar keybindings and faster iteration while writing code.

Yes, even a light weight cell input-output cell that is not as feature rich as euporie would be a pretty neat addition to helix

**Comment by ShalokShalom** [comment]
Couldnt we just integrate euphorie? I know, the core team wants everything to be done in Rust, and still, I see that this could do the job: https://rustpython.github.io/

**Comment by pascalkuthe** [comment]
the issue isn't just that that is not written in rust which is generally a blocker but also that its specific to python. We will not include any language-specific tools or integration in core. That kind ow integration has to wait for blugins

**Comment by ShalokShalom** [comment]
Its not specific to Python. 

There are [dozens of languages](https://github.com/ml-tooling/best-of-jupyter#jupyter-kernels) supported. Also [Rust](https://github.com/evcxr/evcxr/blob/main/evcxr_jupyter/README.md)

**Comment by astro-ray** [comment]
@pascalkuthe 
When considering the comment about euporie, that is very much python oriented, but notebooks are not specific to a single language. Python, R, Julia, all have a popularity of being used with a notebook, particularly in scientific community. I regularly use a combination of Python and Bash in jupyter.

Even Matlab, Wolfram has notebook options

**Comment by pascalkuthe** [comment]
I think including something like https://gitlab.com/Screwtapello/kakoune-repl-buffer can make sense because that interfaces with an arbitrary shell command instead of integrating with one specific REPL. Maype something fancier beyond that may make sense but I don't think we should integrate language specific notebooks/repls into core (and I still count euporie as that, yes it suuports some other languages but its definitly primarily oriented towards python and not universal at all). 

At most we could have an arbitrary REPL mode that can have an API that plugins could provide backends to (with the default backend just being stdout/stdin)

**Comment by huwaireb** [comment]
Just wanted to provide a heads up,
A protocol similar to that of LSP or DAP already exists, has existed for almost a decade now and that is NREPL: https://nrepl.org/nrepl/1.1/index.html. Albeit majorly used in clojure(script) community it is language and editor agnostic as mentioned in the following page: https://nrepl.org/nrepl/beyond_clojure.html, stating and I quote:

> As mentioned earlier the nREPL protocol is language-agnostic and can be leveraged for many languages that have the ability to evaluate code at runtime.

The only odd thing I found in the protocol personally is the usage of bencode for the data transport, but that is merely the default provided option, and the reasoning was provided in the author here: https://stackoverflow.com/questions/59594424/why-bencode-has-been-used-for-transporting-clojure-code-to-nrepl-in-cider.

**Comment by theophilusx** [comment]
From reading some of the comments in this thread, I fear there is a risk of conflating two different user cases/ideas here. One the one side, we have the 'notebook' concept with approaches like Jupyter, euporie and literate style programming environments, similar to Emacs' org-mode etc.  On the other side, you have the repl driven development approach, which is similar, but not the same. The differences are somewhat subtle and not obvious from the outside. You really need to have used both approaches to apprciate these differences. For this reason, I'm not convinced that incorporating something like euporie is appropriate. THe nrepl approach is interesting, but I'm not sure it is correct either. There are other alternatives, such as socket based repls, which are also popular and which apparently address some of the percieved weaknesses with nrepl. I"m also not sure how well nrepl scales to other languages (I've only used it with Clojure/Clojurescript). 

My gut feeling here is that what Helix should really do is take the time to examine and assess how other editors have addressed both the two use cases (i.e. notebookand repl driven development), work out what has been done well and what could be done better and then see if an initial 'common ground' could be found which could be built into Helix as a core feature and which would then be leveraged on to develop more specific functionality for each of the use cases (something which would likely be part of the evenual plugin ecosystem?). 

BTW with respect to repl driven development, I think the neovim conjure plugin may be a good source of ideas. It provides a framework for implementing repls in a number of languages and is able to take advantage of existing interfaces e.g. for Clojure, it communicates over a nrepl connection.  For repl driven development, how it itnerfaces with the repl is quite nice and a model which would liekly work well with Helix (due to also being based around a modal interface).

**Comment by ShalokShalom** [comment]
I can see, that the optional/additional integration by Microsoft - dotnet interactive - can be an advantage here. 

Yes, this would pull dotnet in - but you do not depend on it for normal jupyter support. It is only to add a couple of really cool features, and comes with very little code to pay for:

The main benefit is probably the sharing of [variables between languages.](https://github.com/dotnet/interactive/blob/main/docs/variable-sharing.md) 

![Screenshot_20231211_234612](https://github.com/helix-editor/helix/assets/6344099/2c6f91d1-5cca-42f5-880b-0f2b33ec1bb6)

So Python can bind the result of a calculation to a name, and F# can consume it. 
How to add a language to that support category? 
There is a [really simple guide](https://github.com/dotnet/interactive/blob/main/docs/variable-sharing.md) for it available. 

Please note, that this is ultimately simply another implementation of the Jupyter protocol. 
The only difference is, that the dotnet core can spawn multiple [sub-kernels](https://github.com/dotnet/interactive/blob/main/docs/adding-jupyter-kernels.md#adding-support-for-new-jupyter-subkernels). 

That means, a dotnet core runs a python and a R core, who then can communicate. 
Very useful for data sharing, data analysis and data science. 

Still compatible with regular old standards. 

There is also a neat alternative to .ipynb called .dib, but I forgot what its purpose was, and I am currently struggling to find proper doc to it. But I remind it was a worthwhile consideration as well. https://github.com/dotnet/interactive/issues/3378

Much fun altogether

**Comment by ShalokShalom** [comment]
Also notice, that Microsoft has implemented an improved notebook file format.

I guess it might be a bit out of scope, but it may prove useful.

https://github.com/dotnet/interactive/blob/main/docs/FAQ.md#whats-the-difference-between-a-dib-file-and-an-ipynb-file

**Comment by peter-fm** [comment]
I found the above solutions work for the most part but had trouble with some valid python code causing issues (mainly classes and extra lines within functions). So I wrote a little rust tool to slot into the keyboard shortcut call that fixes most of the indentation issues. If anyone is interested you can find it here: [pypaste](https://github.com/peter-fm/pypaste).

**Comment by ColdTeapot273K** [comment]
Sharing my recipe for Helix + Kitty + virtual environment REPL workflow. Tried to make it simple, robust and straightforward. 

Inspired by [danjenson](https://github.com/danjenson) and [sourproton](https://github.com/sourproton) solutions. 

Prerequisites:
[] Existing virtual environment with dependencies in project root:
  - Python (`Poetry` + `Pyenv`):
      - `pyproject.toml`
      - `poetry.lock`
  - Julia (`Pkg`):
      - `Project.toml`
      - `Manifest.toml`
  
[] `~/.local/bin/ in $PATH

1. Create script:

`~/.local/bin/launch_workspace_repl.sh`

```bash
#!/bin/sh

if [ -f ./pyproject.toml ] && [ -f ./poetry.lock ]; then
    kitty @ launch --type=window --title REPL --keep-focus --cwd current --hold --copy-env sh -c 'pyenv exec poetry run ipython' >/dev/null
elif [ -f ./Project.toml ] && [ -f ./Manifest.toml ]; then
    kitty @ launch --type=window --title REPL --keep-focus --cwd current --hold --copy-env sh -c 'julia -i -e "using Pkg; Pkg.activate(\".\")"' >/dev/null
else 
    echo "Necessary workspace dependencies not found!" 
    exit 1 
fi
```

Make it executable:

`chmod +x ~/.local/bin/launch_workspace_repl.sh`

2. Extend Helix config:

`~/.config/helix/config.toml`

```toml
[keys.normal]
"C-ret" = ":sh launch_workspace_repl.sh"
"S-ret" = [":pipe-to kitty @ send-text --match 'title:REPL' --stdin", ":pipe-to kitty @ send-text --match 'title:REPL' '\r'"]
```

3. When using Helix from project root:
    - Press `Ctrl+Enter` to launch REPL in existing project virtual environment
    - Press `Shift+Enter` to pipe selected code to the REPL

---

Commentary:
- I wish there'd be a way to infer REPL based on active LSP (https://github.com/helix-editor/helix/issues/4501) or at least based on file extension (https://github.com/helix-editor/helix/pull/11164). But inferring from project configs is not the worst workflow, encourages pinning project dependencies.
- Straightforward to extend to other languages and environment mgmt systems.
- In step 1 `--copy-env` and `'pyenv exec poetry run ...` are very specific to make Poetry work.
- In step 2 `":pipe-to kitty @ send-text --match 'title:REPL' '\r'"` is for cases when you want to 
- Could use some [bracketed paste mode](https://cirw.in/blog/bracketed-paste) like in [sourproton](https://github.com/sourproton)'s solution. VSCode Python Interactive has convenience of stopping execution of the whole code selection when errors are thrown (treating the selection like Jupyter cell). On the other hand, the default behaviour of REPLs like IPython and Julia REPL is keep executing next lines of the pasted selection.
- Could use parsing from `%%` cells like in [sourproton](https://github.com/sourproton)'s solution.
- Just in case, tested on:
     - helix = 24.7
     - kitty = 0.35.2
     - poetry =1.8.3
     - pyenv = 2.3.36-1-g8f8a5e0
     - python = 3.11.8
     - julia = 1.10.4+0.x64.linux.gnu
     - os: fedora 40

**Comment by cgahr** [comment]
@sourproton your solution is great!

I slightly refined it using the new macro expansion syntax. I changed the keybindings since they are more ergonomic for me.

My helix config looks as follows:
```toml
[keys.normal]
"[" = "@?^#<space>%%<ret>"
"]" = "@/^#<space>%%<ret>"

[keys.normal.backspace]
s = [":pipe-to tmux -L hx_repl load-buffer - \\; paste-buffer -dpr -t REPL \\; send-keys -t REPL Enter"]
space = [":sh foot tmux -L hx_repl new -n REPL 'ipython || python'"]
t = "@/^#<space>%%<ret>Ni#<esc>nF<ret>;vg.<esc>_<backspace>sun"
```
- the first two commands bind `]b` and `[b` to next and previous block
- `<backspace><space>` runs a new terminal (`foot` in my case but this should work for any terminal), starts tmux, and inside tmux runs ipython
- `<backspace>s` sends the selection to the REPL opened using the previous command
- `<backspace>t` runs the current block in the same REPL

The core idea remains the same, but IMO, there are 2 improvements:
- I use the new macro expansion syntax (which did not exist before). This allows me to run helix outside of tmux
- the REPL doesn't need to be a split pane on the right and I'm much more flexible
- technically it is better to match `# %%` only at the beginning of the line as done here

Maybe this is of interest for someone! And thanks again to @sourproton

**Comment by dhirschfeld** [comment]
> People here seem to be conflating `jupyter` with a notebook environment. `jupyter` itself is a protocol for frontends to talk to kernels (repls) for any of the many languages implemented.
> 
> * [jupyter-protocol.readthedocs.io/en/latest/index.html](https://jupyter-protocol.readthedocs.io/en/latest/index.html)
> 

In case anyone is thinking of hooking Helix up to Jupyter, this is a good technical overview:
* https://www.romaglushko.com/blog/jupyter-kernel-architecture/

**Comment by cgahr** [comment]
Thanks!

> I often end up working with several different projects in parallel, so for me using a hard coded common name for the tmux pane would be an issue. 

I think when #11164 lands, we could use
```toml
[keys.normal.backspace]
s = [":pipe-to tmux -L hx_repl load-buffer - \\; paste-buffer -dpr -t %{basename} \\; send-keys -t REPL Enter"]
space = [":sh foot tmux -L hx_repl new -n %{basename} 'ipython || python'"]
```
and replace `-t REPL` dynamically with the name of the file `-t %{basename}` to get a repl per file (not tested).

**Comment by giovannikraushaar** [comment]
I've achieved something similar with helix + zellij. The downside is that it requires helix to be run inside a zellij session although the startup time difference is negligible. You can do son with a simple wrap script like so this:

```sh
_session_name=helix-$(mktemp -u XXXXX)
_layout_path=$XDG_RUNTIME_DIR/$_session_name.kdl
if [ -n "$1" ]; then
    _args=$(printf '\n\t\t\targs %s' "$(printf '"%s" ' "$@")")
fi
printf 'session_name "%s"\nlayout {\n\tdefault_tab_template {\n\t\tchildren\n\t}\n\ttab name="editor" {\n\t\tpane name="Editor" {\n\t\t\tcommand "/usr/bin/hx"%s\n\t\t\tborderless true\n\t\t\tclose_on_exit true\n\t\t}\n\t}\n}\ndefault_mode "locked"\nsession_serialization false' "$_session_name" "$_args" > "$_layout_path"
/usr/bin/zellij --new-session-with-layout "$_layout_path"
rm "$_layout_path"
```

---

The command is basicalliy:

```toml
C-ret = ":pipe-to ~/.config/helix/script/send-to-session.sh %{language}"
```

where `~/.config/helix/script/send-to-session.sh` is the following:

```sh
#! /usr/bin/dash
#
# Dependencies: helix zellij ripgrep (optional:) git R ipython fish
# Usage: [FROM HELIX] :pipe-to ~/.config/helix/script/send-to-session.sh %{language}


if [ -z "$1" ]; then
  echo "Missing mandatory argument language"
  exit 1
fi

if [ -z "$ZELLIJ" ]; then
  echo "Not running in a zellij session"
  exit 3
fi

_TEXT=$(cat)
_LANG="$1"
_CWD=$(if git -v > /dev/null 2>&1; then if git rev-parse --show-toplevel > /dev/null 2>&1; then git rev-parse --show-toplevel; else pwd; fi; else pwd; fi)
_LAYOUT_FILE="$XDG_RUNTIME_DIR/REPL-$(mktemp -u XXXXX).kdl"

case "$_LANG" in
  r)
    _TAB=r-console
    _CMD=/usr/bin/R 
    _ARGS='args "--quiet" "--no-save" "--no-restore-data" "--interactive"'
    ;;
  python)
    _TAB=py-console
    _CMD=/usr/bin/ipython
    _ARGS='args "--quick" "--no-autoindent" "--pprint" "--no-banner" "--no-confirm-exit"'
    ;;
  fish)
    _TAB=fish-console
    _CMD=/usr/bin/fish
    ;;
  bash)
    _TAB=bash-console
    _CMD=/usr/bin/bash
    ;;
  *)
    echo "Unsupported language $1"
    exit 4
    ;;
esac

if [ -z "$(zellij action query-tab-names | rg $_TAB)" ]; then
cat << EOF > "$_LAYOUT_FILE"
attach_to_session true
layout {
  tab_template name="console" {
    children
    pane size=1 borderless=true split_direction="horizontal" {
      plugin location="zellij:compact-bar"
    }
  }
  console name="$_TAB" split_direction="vertical" {
    pane name="REPL" {
      command "$_CMD"
      $_ARGS
      cwd "$_CWD"
      focus true
      close_on_exit true
    }
  }
}
EOF
  zellij action new-tab --name "$_TAB" --layout "$_LAYOUT_FILE" --cwd "$_CWD"
else
  zellij action go-to-tab-name "$_TAB"
fi
zellij action write-chars "$_TEXT" && \
zellij action write 13 && \  # send enter key
exit 0
```

I use pipe-to to properly escape any control sequence inside the selected text and exploit the power of the new variable substitution to select the correct REPL.

**Comment by j3ka** [comment]
I've created little app for such purpose: [https://github.com/j3ka/reple](https://github.com/j3ka/reple). It creates wrapper over real repl, and then uses named pipe to transfer your block of code to repl. You can use it without zellij, tmux, screen, etc. You can even use separate terminal window for this.
There are screenshots, how to use it with helix and `pipe-to`:

<img width="1360" height="516" alt="Image" src="https://github.com/user-attachments/assets/e4948250-8aad-4607-b00e-cd571bd9621a" />
<img width="1360" height="516" alt="Image" src="https://github.com/user-attachments/assets/e1bbe6af-5002-4b9d-ab87-efbd6475b8ea" />

Of course I have keybindings like:
```toml
[keys.normal]
"A-ret" = ":pipe-to reple eval"

[keys.select]
"A-ret" = ":pipe-to reple eval"

```

PS: this is not a promotion of my application. just sharing a solution.

**Comment by peter-fm** [comment]
Nice, thanks for sharing @j3ka! I tested it and it works with [pypaste](https://github.com/peter-fm/pypaste). e.g, `"A-ret" = ":pipe-to pypaste | reple eval"`

**Comment by rawhuul** [comment]
Hi, I see as there are few workaround to get REPL behaviour but is this feature planned?, as I can’t use current workarounds since they rely on TMUX and I’m windows user.

**Comment by jerabaul29** [comment]
I guess with the plugin system landing soon, there will probabably soon be some form of plugin for this emerging? :) But I must say that on Linux I have grown very found of the tmux (or similar) solutions described above :) .

**Comment by rawhuul** [comment]
> I guess with the plugin system landing soon, there will probabably soon be some form of plugin for this emerging? :) But I must say that on Linux I have grown very found of the tmux (or similar) solutions described above :) .

I really appreciate that, waiting for plugin system to roll out

**Comment by ColdTeapot273K** [comment]
For selecting text between `# %%` tokens denoting cells I wish there was "match inside regex pattern" command for [textobject options](https://docs.helix-editor.com/textobjects.html)

e.g. something like `mir<regex>`

Alternatively `search for regex pattern` could have been used. 
But underlying rust regex implementation [doesn't support negative lookahead](https://github.com/rust-lang/regex/issues/127). So something like `^(?!# %%).*` from python regex doesn't work.

**Comment by timothymillar** [comment]
Just to add a simple vim-slime inspired config with tmux:
```
[keys.select.z]
# runs the selection in the 'next' tmux pane and returns to normal mode
z = [
    ''':pipe-to tmux load-buffer - \; paste-buffer -dpr -t '{next}' \; send-keys -t '{next}' Enter''',
    '''normal_mode'''
]

[keys.normal.z]
# selects and runs the current paragraph
z = "@mipvzz"
# selects and runs the current line
x = "@vxzz"
```
which lets me hit `zz` or `zx` to run the current paragraph or line respectively.

I also wanted an option to 'compile' a percent delimited script into a markdown report with figures using [jupytext](https://jupytext.readthedocs.io/en/latest/index.html) and [nbconvert](https://nbconvert.readthedocs.io/en/latest/):
```
[keys.normal.z]
# compile markdown report with jupytext and nbconvert 
j = ":sh jupytext %{buffer_name} --to notebook --execute -o _tmp.ipynb && jupyter nbconvert _tmp.ipynb --to markdown --output $(basename %{buffer_name} .py).md && rm _tmp.ipynb"
```

**Comment by waddie** [comment]
It’s not universal, but I hacked together an nREPL plugin over the weekend: https://github.com/waddie/nrepl.hx

![nrepl.hx demo with Clojure](https://github.com/waddie/nrepl.hx/blob/main/images/nrepl.gif?raw=true)

Currently has a Rust dependency, but I’m pretty sure that can all be migrated to Steel eventually.

I’ve only built it on my machine (macOS Tahoe) and mostly been testing with Clojure and Babashka, but I did get it working with [nrepl-python](https://git.sr.ht/~ngraves/nrepl-python) just now.

If anyone uses any other language’s server (https://nrepl.org/nrepl/beyond_clojure.html), I’d love to hear if it works there. Or just if you’re using it all, and I’ll maybe stop yolo-ing every thought that crosses my mind into `main`. 😄

**Comment by wolf** [comment]
I built something that might be relevant here: [dap-mux](https://github.com/dap-mux/dap-mux), a DAP multiplexer.

It sits between Helix and a debug adapter and lets multiple clients share the same debug session simultaneously. The practical result: Helix handles source tracking and breakpoints while an IPython REPL handles stepping and expression evaluation — both connected to the same session at the same time.

From Helix's side it's just `:debug-remote host:port attach`. No changes to Helix required.

There's a side effect worth mentioning: since the debug adapter is dap-mux's child process rather than Helix's, the terminal corruption bug (#6265) goes away entirely.

First public release — tested live with Helix + debugpy + IPython. If this is what people here have been looking for, bug reports and notes from other configurations are very welcome.
